public/app/features/dashboard/components/GenAI/hooks.ts

```diff
@@ -0,0 +1,95 @@
+import { useState } from 'react';
+import { useAsync } from 'react-use';
+import { Subscription, finalize } from 'rxjs';
+
+import { llms } from '@grafana/experimental';
+
+// TODO: Add tests
+export function useOpenaiStreamHook(
+  model = 'gpt-3.5-turbo',
+  temperature = 1
+): {
+  setMessages: React.Dispatch<React.SetStateAction<llms.openai.Message[]>>;
+  reply: string;
+  started: boolean;
+  finished: boolean;
+  loading: boolean;
+  error: Error | undefined;
+  value:
+    | {
+        enabled: any;
+        stream?: undefined;
+      }
+    | {
+        enabled: any;
+        stream: Subscription;
+      }
+    | undefined;
+} {
+  // The messages array to send to the LLM, updated when the button is clicked.
+  const [messages, setMessages] = useState<llms.openai.Message[]>([]);
+  // The latest reply from the LLM.
+  const [reply, setReply] = useState('');
+
+  const [started, setStarted] = useState(false);
+  const [finished, setFinished] = useState(true);
+
+  const { loading, error, value } = useAsync(async () => {
+    // Check if the LLM plugin is enabled and configured.
+    // If not, we won't be able to make requests, so return early.
+    const enabled = await llms.openai.enabled();
+    if (!enabled) {
+      return { enabled };
+    }
+    if (messages.length === 0) {
+      return { enabled };
+    }
+
+    setStarted(true);
+    setFinished(false);
```

Comment:
Can't this just be `setInProgress`?

Reply:
I tried this and it made state management significantly messier, in my opinion. e.g., I ended up wanting to set a hasAlreadyRun state (inProgress has never been true), and then putting branching logic on effects with inProgress as a dependency.

Reply:
I'll take a stab at implementing this 👍